hy: 1.0a4 -> 0.24.0

thiagokokada opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Not a downgrade, upstream returned to 0.* version numbers: https://github.com/hylang/hy/releases/tag/0.24.0. `python3Packages.hyrule` was updated in one of the mass Python upgrades.

Also updated `funcparserlib` since `hy` explicitly depends on this specific version (it is the latest, BTW). We could relax the deps instead, but in this case bumping the library was easier (and also removed a patch).

Fixes https://github.com/NixOS/nixpkgs/issues/194527.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

thiagokokada committed:
funcparserlib: 1.0.0a0 -> 1.0.0

thiagokokada committed:
hy: 1.0a4 -> 0.24.0

Not a downgrade, upstream returned to 0.* version numbers:
https://github.com/hylang/hy/releases/tag/0.24.0

thiagokokada commented:
CC @fabaff.

thiagokokada commented:
Result of `nixpkgs-review pr 194558` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>23 packages built:</summary>
  <ul>
    <li>actdiag (python310Packages.actdiag)</li>
    <li>asciidoc-full-with-plugins</li>
    <li>blockdiag (python310Packages.blockdiag)</li>
    <li>hy (python310Packages.hy)</li>
    <li>nwdiag (python310Packages.nwdiag)</li>
    <li>python310Packages.funcparserlib</li>
    <li>python310Packages.hyrule</li>
    <li>seqdiag (python310Packages.seqdiag)</li>
    <li>python310Packages.sphinxcontrib-actdiag</li>
    <li>python310Packages.sphinxcontrib-blockdiag</li>
    <li>python310Packages.sphinxcontrib-nwdiag</li>
    <li>python310Packages.sphinxcontrib-seqdiag</li>
    <li>python39Packages.actdiag</li>
    <li>python39Packages.blockdiag</li>
    <li>python39Packages.funcparserlib</li>
    <li>python39Packages.hy</li>
    <li>python39Packages.hyrule</li>
    <li>python39Packages.nwdiag</li>
    <li>python39Packages.seqdiag</li>
    <li>python39Packages.sphinxcontrib-actdiag</li>
    <li>python39Packages.sphinxcontrib-blockdiag</li>
    <li>python39Packages.sphinxcontrib-nwdiag</li>
    <li>python39Packages.sphinxcontrib-seqdiag</li>
  </ul>
</details>

thiagokokada commented:
CC @shadowrylander since you opened #194527.

shadowrylander commented:
Thanks so much! I've been trying to get `withPackages` to work with my upgraded overlay, but if I tried to run the `hy` repl through a flake with a `hy` package, it would tell me it can't find the binary; I was hoping I'd have more luck with the official version!

thiagokokada committed:
hy: fix passthru.withPackages

In the previous version it was adding the extra packages as
`propagatedBuildInputs`. This meant that this package would be rebuild
for no reason (it is not like the package will actually depend on the
extra inputs) and also cause the strange side-effect of creating a hy
package without its console entry-points (e.g.: no `$out/bin` contents).

Now we are reusing the `python.withPackages` that will avoid the
unnecessary rebuild, fixing both issues.

thiagokokada commented:
> Thanks so much! I've been trying to get `withPackages` to work with my upgraded overlay, but if I tried to run the `hy` repl through a flake with a `hy` package, it would tell me it can't find the binary; I was hoping I'd have more luck with the official version!

It seems that there was an issue on how `hy.withPackages` was defined before. I rewrote the function and it should work now.

Can you try again?

shadowrylander commented:
There's a slight issue with this, in that it runs the _Python_ repl, not `hy`, though still with the packages.

thiagokokada commented:
> There's a slight issue with this, in that it runs the _Python_ repl, not `hy`, though still with the packages.

What do you mean? `hy` binary should be available since I explicitly inject it to the environment:

```nix
{ pkgs ? import /home/thiagoko/Projects/nixpkgs { } }:

with pkgs;
mkShell {
  buildInputs = [
    (python39Packages.hy.withPackages (ps: with ps; [ hyrule requests ]))
  ];
}
```

```
✦ ~/tmp 6s
❯ nix-shell

~/tmp
nix-shell ❯ hy                                                                                                                                                                                                                             nix-shell
Hy 0.24.0 using CPython(main) 3.9.14 on Linux
=>
```

shadowrylander commented:
Yes, it's in the environment, but not available using `nix run` when using packages with `hy` as a dependency. Is there any way to fix that?

thiagokokada commented:
> Yes, it's in the environment, but not available using `nix run` when using packages with `hy` as a dependency. Is there any way to fix that?

Not that I know and also it is not why `withPackages` exist.

thiagokokada commented:
@shadowrylander BTW, can you review the package? I think whatever you want should be created as an issue with a clear use case.

Issue #194527 was related to the bump of this package that this PR should solve already.

shadowrylander commented:
Right; makes sense. So by review, do you mean just look through the changes, or something else? Sorry, I'm a little new to this!

thiagokokada commented:
> Right; makes sense. So by review, do you mean just look through the changes, or something else? Sorry, I'm a little new to this!

Yeah. Would be nice if you could also test the package and make sure that everything is still working as expected (except for your specific issue not related to the bump).

> Also, may I tag you in the new issue?

Sure. I can't promise anything though.

shadowrylander commented:
In that case, yep! Everything works as expected; thanks for all the help! And I have tagged you in the new issue!
